README.md
# Euler - Mathematics functions for Haskell.

[![Build Status](https://travis-ci.org/luisonthekeyboard/euler.svg?branch=master)](https://travis-ci.org/luisonthekeyboard/euler)

## What is this?

This repo contains functions to help in mathematics calculations or puzzles using Haskell. Currently there is only one module, `Numeric.Euler.Primes`, which contains implementations of three prime number sieves:

- [Sieve of Erastothenes](https://en.wikipedia.org/wiki/Sieve_of_Eratosthenes)
- [Sieve of Sundaram](https://en.wikipedia.org/wiki/Sieve_of_Sundaram)
- [Sieve of Atkin](https://en.wikipedia.org/wiki/Sieve_of_Atkin)

The current major version (0) is stable enough to be used for _some_ Project Euler problems but keep in mind that the sieves are not optimized at all.

You can also find this package [in Hackage](http://hackage.haskell.org/package/euler).


## Dependencies and building

To install the library you just need the latest `base`. Just `cabal install euler` from you project or add `euler` as a dependency to your `.cabal` file and you should be good to go.

If you want to build the library locally, you'll need a few more things:


### Building with `cabal`

You'll need:

- `GHC >= 8.6`
- `cabal-install >= 3.*`

`cabal` will take care of installing all the other dependencies.

    git clone https://github.com/luisonthekeyboard/euler.git
    cd euler
    cabal build
    cabal test


Happy hacking!
:-)


changelog.md
Change Log
==========

0.10.1 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2020, May 17th
* Update project to Cabal 3
* Remove performance testing, test coverage, benchmarks, runhaskel support and -Wall

---

0.9.2 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 205, Apr 16th
* Removes logging information.
* Builds with GHC 7.10.

---

0.8.1 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 19th
* Logs more information while running.
* Documentation update.

---

0.8.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 15th
* Adds examples of build scripts that use both `cabal` and `runhaskell`;
* Makes the HPC script more resilient to a build done with `cabal` vs a build done with `runhaskell`;
* Removes dependency on `cabal` for the haddock phase.

---

0.7.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 13th
* Fixes for HPC test-suite.

---

0.6.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 7th
* Makes happy a required build tool;
* Fixes GHC warnings.

---

0.5.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 7th
* Adds `changelog.md`;
* Adds Criterion benchmarks;
* Adds support for travis build.

---

0.4.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 6th
* Adds full documentation to the Primes module;
* Made the default haddock threshold become 100%;
* Brought down the limit for prime number generation during spec test phase;
* Removed temporary files;
* Updated dependencies;
* Changed formatting;
* Made script to run documentation;
* Makes quick check tests run more cases;
* Updated build phases;
* Full test coverage;
* Adds quickcheck tests as well as HPC coverage.

---

0.3.3 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Mar 3rd
* Fixed upper bounds and lower bounds in dependencies;
* Fixed all HLint suggestions;
* Added HLint to the build;
* Fixed all -Wall errors on test phase;
* Fixed all -Wall errors on compilation.

---

0.3.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Feb 24th
* Made -Wall default compilation option
* Added source-repository field to cabal file.

---

0.2.2 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Feb 23rd
* Added a unit test;
* Removed warnings from cabal file; Updated references to dependencies;
* Fixed errors all over the place;
* Removed unused functions; cleaned up indentation.

---

0.1.0 Luis Rodrigues Soares <luis@luisonthekeyboard.com> 2015, Feb 21st
* Added description to package;
* Fixed module references;
* Cabalized the project;
* Initial copy of Problem07.hs, from https://github.com/luisonthekeyboard/projectEuler


test/HPC.hs
module Main (main) where

import Data.List
import System.Exit
import Text.XML.Light
import Data.Maybe
import System.Process

expected :: Fractional a => a
expected = 85

simpleQName :: String -> QName
simpleQName name = QName{qName=name, qURI=Nothing, qPrefix=Nothing}

paramsForCabal :: String -> [String]
paramsForCabal tixFile = ["report", tixFile, "--include=Numeric.Euler.Primes"]

usefulMetrics :: QName -> Bool
usefulMetrics name = name `elem` [
  simpleQName "exprs",
  simpleQName "booleans",
  simpleQName "alts",
  simpleQName "local",
  simpleQName "toplevel"]

extractValues :: Element -> (String,String)
extractValues element = (
  fromMaybe ("no string"::String) (findAttrBy (\name -> name == simpleQName "boxes") element),
  fromMaybe ("no string"::String) (findAttrBy (\name -> name == simpleQName "count") element))

calculatePercentages :: (String,String) -> Float
calculatePercentages tuple =
  100 *
  (read (snd tuple) :: Float) /
  (read (fst tuple) :: Float)

coverageOk :: [Float] -> Bool
coverageOk values = (realToFrac (sum values) / genericLength values) > (expected :: Double)

hpc :: [String] -> IO String
hpc tixFile = readProcess "hpc" (tixFile ++ ["--xml-output"]) ""

readCoverageFrom :: [String] -> IO ()
readCoverageFrom tixFile = do
  stdout <- hpc tixFile
  case parseXMLDoc stdout of 
   Nothing -> error "Failed to parse xml. Try running HPC manually..."
   Just doc -> let elements = filterChildrenName usefulMetrics (head $ elChildren doc)
                   values = map extractValues elements
                   percents = map calculatePercentages values
               in if coverageOk percents
                  then exitSuccess
                  else
                    do
                      humanReadableOutput <- readProcess "hpc" tixFile ""
                      putStrLn ""
                      putStrLn humanReadableOutput
                      exitFailure

main :: IO ()
main = readCoverageFrom (paramsForCabal "./spec.tix")


test/haddock.hs
module Main (main) where

import Data.List
import System.Exit
import System.Process
import Text.Regex.Posix

expected :: Fractional a => a
expected = 100

readMatchAsInt :: [String] -> Float
readMatchAsInt list = read (list !! 1) :: Float

coverageOk :: [Float] -> Bool
coverageOk values = (realToFrac (sum values) / genericLength values) >= (expected :: Double)
--coverageOk = all (> expected)

main :: IO ()
main = do
  output <- readProcess "haddock" ["src/Numeric/Euler/Primes.hs"] ""
  let percents = map readMatchAsInt (output =~ "^ *([0-9]*)% " :: [[String]]) in
   if coverageOk percents
   then exitSuccess
   else putStr output >> exitFailure



test/EulerSpec.hs
module EulerSpec (main, spec) where

import Test.Hspec
import Euler

main :: IO ()
main = hspec spec

spec :: Spec
spec = do
  
  describe "isPrime" $ do
    it "works" $ do
      isPrime2 7 `shouldBe` True
--      head [23 ..] `shouldBe` (23 :: Int)

--    it "returns the first element of an *arbitrary* list" $
--    property $ \x xs -> head (x:xs) == (x :: Int)

--    it "throws an exception if used with an empty list" $ do
--    evaluate (head []) `shouldThrow` anyException


src/Euler.hs
module Euler (isPrime2) where

import Data.List (sort)


{-------------------------------------------------
-------------------------------------------------}


isPrime2 :: Integral a => a -> Bool
isPrime2 n
  | n == 2    = True
  | otherwise = not (any (\i -> n `mod` i == 0) [3..n-1])

trialAndDivision2 limit = 2:3:5:[n | n <- [7,9..limit], isPrime2 n]


{-------------------------------------------------
-------------------------------------------------}


erastothenes3 limit = eSieve3 [3,5..limit] [2]

eSieve3 [] primes     = primes
eSieve3 (x:xs) primes = eSieve3 ([y | y <- xs, y `mod` x /= 0]) (x:primes)


{-------------------------------------------------
-------------------------------------------------}

initialSundaramSieve :: Int -> [Int]
initialSundaramSieve limit =
  let topi = floor (sqrt ((fromIntegral limit) / 2)) in
  [i + j + 2 * i * j | i <- [1..topi],
   j <- [i..floor((fromIntegral(limit-i)) / fromIntegral(2*i+1))]]

sundaram5 limit =
  let halfLimit = floor((limit / 2)-1) in
  2:removeComposites ([1..halfLimit]) (sort $ initialSundaramSieve halfLimit) 

removeComposites [] _ = []
removeComposites (n:ns) [] = (2*n+1) : (removeComposites ns [])
removeComposites (n:ns) (c:cs)
  | n == c    = removeComposites ns cs
  | n > c     = removeComposites (n:ns) cs
  | otherwise = (2*n+1) : (removeComposites ns (c:cs))

{---------------------------------------------------
---------------------------------------------------}


initialAtkinSieve :: Int -> [(Int,Int)]
initialAtkinSieve limit =
  sort $ zip [n
             | n <- [60 * w + x
                    | w <- [0..limit `div` 60],
                      x <- [1,7,11,13,17,19,23,29,31,37,41,43,47,49,53,59]
                    ],
               n <= limit]
  (take limit [0,0..])


aFlip :: (Int,Int) -> (Int,Int)
aFlip (x,1) = (x,0)
aFlip (x,0) = (x,1)


flipAll :: [Int] -> [(Int, Int)] -> [(Int, Int)]
flipAll _      [] = []
flipAll []     ss = ss
flipAll (f:fs) ((s,b):ss)
  | s < f      = (s,b): (flipAll (f:fs) ss)
  | s == f     = flipAll fs ((aFlip (s,b)):ss)
  | otherwise  = flipAll fs ((s,b):ss)


firstStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
firstStep size sieve =
  let topx = floor(sqrt(fromIntegral (size `div` 4)))
      topy = floor(sqrt(fromIntegral size)) in
  flipAll (sort [n | n <- [4*x^2 + y^2| x <- [1..topx],
                           y <- [1,3..topy]],
                 n `mod` 60 `elem` [1,13,17,29,37,41,49,53]])
  sieve


secondStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
secondStep size sieve =
  let topx = floor(sqrt(fromIntegral (size `div` 3)))
      topy = floor(sqrt(fromIntegral size)) in
  flipAll (sort [n | n <- [3*x^2 + y^2 | x <- [1,3..topx],
                           y <- [2,4..topy]],
                 n `mod` 60 `elem` [7,19,31,43]])
  sieve


thirdStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
thirdStep size sieve =
  let topx = floor(sqrt(fromIntegral size)) in
  flipAll (sort [n | n <- [3*x^2 - y^2 | x <- [1..topx],
                           y <- [(x-1),(x-3)..1],
                           x > y],
                 n `mod` 60 `elem` [11,23,47,59]])
  sieve


unmarkMultiples limit n sieve =
  let nonPrimes = [y | y <-[n,n+n..limit]] in
  unmarkAll nonPrimes sieve


unmarkAll _        []    = []
unmarkAll []       sieve = sieve
unmarkAll (np:nps) ((s,b):ss) 
  | np == s   = unmarkAll nps ss
  | np < s    = unmarkAll nps ((s,b):ss)
  | otherwise = (s,b) : (unmarkAll (np:nps) ss)


atkin1 :: Int ->  [(Int,Int)]
atkin1 limit =
  aSieve1 limit ((thirdStep limit) . (secondStep limit) . (firstStep limit) $ initialAtkinSieve limit) [(5,1),(3,1),(2,1)]


aSieve1 _     []         primes = primes
aSieve1 limit ((x,b):xs) primes
  | b == 1 = aSieve1 limit (unmarkMultiples limit (x^2) xs) ((x,b): primes)
  | otherwise = aSieve1 limit xs primes


src/Numeric/Euler/Primes.hs
{-|
Module      : Primes
Description : Module with functions to calculate Primme Numbers. Implements 3
              sieves.
Copyright   : (c) Luis Rodrigues Soares, 2015
License     : MIT
Maintainer  : luis@decomputed.com
Stability   : experimental
-}

module Numeric.Euler.Primes (
  isPrime,
  trialAndDivision,
  erastothenes,
  sundaram,
  atkin
  ) where

import Data.List (sort)


{-------------------------------------------------
-------------------------------------------------}

{-|
The `isPrime` function checks whether a number `a` is prime by successively
checking the remainder of the integer division with all numbers up to @a-1@.hs
-}
isPrime :: Integral a => a -> Bool
isPrime n
  | n <  2    = False
  | n == 2    = True
  | otherwise = not (any (\i -> n `mod` i == 0) [2..n-1])

{-|
The `trialAndDivision` function calculates prime numbers up to a certain limit
by using the traditional (and very inneficient) trial and division method.
-}
trialAndDivision :: Int -> [Int]
trialAndDivision limit = 2:3:5:[n | n <- [7,9..limit], isPrime n]


{-------------------------------------------------
-------------------------------------------------}

{-|
This is a naïve implementation of the Sieve of Erastothenes.
-}
erastothenes :: Int -> [Int]
erastothenes limit = eSieve [3,5..limit] [2]

eSieve :: [Int] -> [Int] -> [Int]
eSieve [] primes     = primes
eSieve (x:xs) primes = eSieve [y | y <- xs, y `mod` x /= 0] (x:primes)


{-------------------------------------------------
-------------------------------------------------}

initialSundaramSieve :: Int -> [Int]
initialSundaramSieve limit =
  let topi = floor (sqrt (fromIntegral limit / 2) :: Double) in
  [i + j + 2 * i * j | i <- [1..topi],
   j <- [i..floor ((fromIntegral(limit-i) / fromIntegral(2*i+1)) :: Double)]]

{-|
An implementation of the Sieve of Sundaram.
-}
sundaram :: Int -> [Int]
sundaram limit =
  let halfLimit = floor( ((fromIntegral limit / 2)-1) :: Double) in
  2:removeComposites [1..halfLimit] (sort $ initialSundaramSieve halfLimit)

removeComposites :: [Int] -> [Int] -> [Int]
removeComposites [] _ = []
removeComposites (n:ns) [] = (2*n+1) : removeComposites ns []
removeComposites (n:ns) (c:cs)
  | n == c    = removeComposites ns cs
  | n > c     = removeComposites (n:ns) cs
  | otherwise = (2*n+1) : removeComposites ns (c:cs)

{---------------------------------------------------
---------------------------------------------------}


initialAtkinSieve :: Int -> [(Int,Int)]
initialAtkinSieve limit =
  sort $ zip [n
             | n <- [60 * w + x
                    | w <- [0..limit `div` 60],
                      x <- [1,7,11,13,17,19,23,29,31,37,41,43,47,49,53,59]
                    ],
               n <= limit]
  (take limit [0,0..])


aFlip :: (Int,Int) -> (Int,Int)
aFlip (x,y)
  | y == 0        = (x,1)
  | y == 1        = (x,0)
  | otherwise     = error "Needs zero or one"


flipAll :: [Int] -> [(Int, Int)] -> [(Int, Int)]
flipAll _      [] = []
flipAll []     ss = ss
flipAll (f:fs) ((s,b):ss)
  | s < f      = (s,b): flipAll (f:fs) ss
  | s == f     = flipAll fs (aFlip (s,b):ss)
  | otherwise  = flipAll fs ((s,b):ss)


firstStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
firstStep size sieve =
  let topx = floor( sqrt(fromIntegral (size `div` 4)) :: Double)
      topy = floor( sqrt(fromIntegral size) :: Double) in
  flipAll (sort [n | n <- [4*x^(2::Integer) + y^(2::Integer)| x <- [1..topx],
                           y <- [1,3..topy]],
                 n `mod` 60 `elem` [1,13,17,29,37,41,49,53]])
  sieve


secondStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
secondStep size sieve =
  let topx = floor( sqrt(fromIntegral (size `div` 3)) :: Double)
      topy = floor( sqrt(fromIntegral size) :: Double) in
  flipAll (sort [n | n <- [3*x^(2::Integer) + y^(2::Integer) | x <- [1,3..topx],
                           y <- [2,4..topy]],
                 n `mod` 60 `elem` [7,19,31,43]])
  sieve


thirdStep :: Int ->  [(Int,Int)] ->  [(Int,Int)]
thirdStep size sieve =
  let topx = floor( sqrt(fromIntegral size) :: Double) in
  flipAll (sort [n | n <- [3*x^(2::Integer) - y^(2 :: Integer) | x <- [1..topx],
                           y <- [(x-1),(x-3)..1]],
                 n `mod` 60 `elem` [11,23,47,59]])
  sieve

unmarkMultiples :: Int -> Int -> [(Int,Int)] -> [(Int,Int)]
unmarkMultiples limit n sieve =
  let nonPrimes = [n,n+n..limit] in
  unmarkAll nonPrimes sieve


unmarkAll :: [Int] -> [(Int,Int)] -> [(Int, Int)]
unmarkAll _        []    = []
unmarkAll []       sieve = sieve
unmarkAll (np:nps) ((s,b):ss)
  | np == s   = unmarkAll nps ss
  | np < s    = unmarkAll nps ((s,b):ss)
  | otherwise = (s,b) : unmarkAll (np:nps) ss


{-|
An implementation of the sieve of Atkin.
-}
atkin :: Int ->  [(Int,Int)]
atkin limit =
  aSieve limit (thirdStep limit .
                secondStep limit .
                firstStep limit $
                initialAtkinSieve limit) [(5,1),(3,1),(2,1)]


aSieve :: Int -> [(Int,Int)] ->[(Int,Int)] -> [(Int,Int)]
aSieve _     []         primes = primes
aSieve limit ((x,b):xs) primes
  | b == 1 = aSieve limit (unmarkMultiples limit (x^(2::Integer)) xs) ((x,b): primes)
  | otherwise = aSieve limit xs primes


euler.hs
module Euler where

import Data.List

--
-- Trial and division
-- yaaaaaaaaaaaaaaaaay
--


isPrime :: Integral a => a -> Bool
isPrime n | n == 2		 = True
	  | n < 2 || even n    	 = False
	  | otherwise 	  	 = not (any (\i -> n `mod` i == 0) [3..n-1])

trialAndDivision limit = [n | n <- [2..limit], isPrime n]


----------------------

isPrime2 :: Integral a => a -> Bool
isPrime2 n | n == 2		 = True
	  | otherwise 	  	 = not (any (\i -> n `mod` i == 0) [3..n-1])

trialAndDivision2 limit = 2:3:5:[n | n <- [7,9..limit], isPrime2 n]


--
-- Sieves
-- Contains implementations of several sieves
--



{------------------------------------------------
-------------------------------------------------}

erastothenes1 limit = eSieve1 [2..limit] []

eSieve1 [] primes = primes
eSieve1 (x:xs) primes = eSieve1
			([y | y <- xs, not (y `isMultiple` x)])
			(primes++[x])

isMultiple :: Int -> Int -> Bool
isMultiple n = \x -> elem n [x,x+x..n]

{------------------------------------------------
-------------------------------------------------}

erastothenes2 limit = eSieve2 [2..limit] []

eSieve2 [] primes = primes
eSieve2 (x:xs) primes = eSieve2
			([y | y <- xs, y `mod` x /= 0])
			(primes++[x])

{------------------------------------------------
-------------------------------------------------}

erastothenes3 limit = eSieve3 [3,5..limit] [2]

eSieve3 [] primes = primes
eSieve3 (x:xs) primes = eSieve3
			([y | y <- xs, y `mod` x /= 0])
			(x:primes)

{------------------------------------------------
-------------------------------------------------}

sundaram1 limit = let limit' = (limit `div` 2) in
	  	      2 : (map (\n -> 2*n+1)
	       	     	  ([1..limit'] \\ [i + j + 2*i*j | i <- [1..limit'], j <- [i..limit']]))

{------------------------------------------------
-------------------------------------------------}

sundaram2 limit = let limit' = (limit `div` 2) in
	  	      2:[2*n+1 | n <- [1..limit'],
		      	       	 not (n `elem` [i + j + 2*i*j | i <- [1..limit'],
				     	       	      	      	j <- [i..limit']])]

{------------------------------------------------
-------------------------------------------------}

sundaram3 limit = let limit' = (limit `div` 2) 
	  	      sieve = [i + j + 2*i*j | i <- [1..limit'],
				     	       j <- [i..limit']] in
	  	  2:[2*n+1 | n <- [1..limit'],
		      	     not (n `elem` sieve)]

{------------------------------------------------
-------------------------------------------------}

sundaram4 :: Int -> [Int]
sundaram4 limit = let limit' = (limit `div` 2) in
	  	  let sieve = [i + j + 2*i*j | let n' = fromIntegral limit',
                           	       	       i <- [1..floor (sqrt (n' / 2))],
                           		       let i' = fromIntegral i,
                           		       j <- [i..floor( (n'-i')/(2*i'+1))]] in
	  	  2:[2*n+1 | n <- [1..limit'],
		      	     not (n `elem` sieve)]

{------------------------------------------------
-------------------------------------------------}

initialSundaramSieve limit = let topi = floor (sqrt ((fromIntegral limit) / 2)) in
	  	             [i + j + 2*i*j | i <- [1..topi],
                           		      j <- [i..floor((fromIntegral(limit-i)) / fromIntegral(2*i+1))]]

sundaram5 limit = let halfLimit = floor((limit / 2)-1) in
	  	      2:removeComposites ([1..halfLimit]) (sort $ initialSundaramSieve halfLimit) 

removeComposites []     _                  = []
removeComposites (n:ns) []                 = (2*n+1) : (removeComposites ns [])
removeComposites (n:ns) (c:cs) | n == c    = removeComposites ns cs
		 	       | n > c     = removeComposites (n:ns) cs
			       | otherwise = (2*n+1) : (removeComposites ns (c:cs))

{--------------------------------------------------
----------------------------------------------------}

initialAtkinSieve limit = sort $ zip
				  [n | n <- [60 * w + x | w <- [0..limit `div` 60],
				      	                  x <- [1,7,11,13,17,19,23,29,31,37,41,43,47,49,53,59]],
					    n <= limit]							
				  (take limit [0,0..])


aFlip :: (x,Int) -> (x,Int)
aFlip (x,1) = (x,0)
aFlip (x,0) = (x,1)


flipAll :: [Int] -> [(Int, Int)] -> [(Int, Int)]
flipAll _      []     = []
flipAll []     ss     = ss
flipAll (f:fs) ((s,b):ss) 
	| s < f		  = (s,b): (flipAll (f:fs) ss)
	| s == f	  = flipAll fs ((aFlip (s,b)):ss)
	| otherwise	  = flipAll fs ((s,b):ss)


firstStep size sieve =
	  let topx = floor(sqrt(fromIntegral (size `div` 4)))
	      topy = floor(sqrt(fromIntegral size)) in
	  flipAll (sort [n | n <- [4*x^2 + y^2| x <- [1..topx],
		  	  	       	  	y <- [1,3..topy]],
			     n `mod` 60 `elem` [1,13,17,29,37,41,49,53]])
                   sieve


secondStep size sieve =
	   let topx = floor(sqrt(fromIntegral (size `div` 3)))
	       topy = floor(sqrt(fromIntegral size)) in
           flipAll (sort [n | n <- [3*x^2 + y^2 | x <- [1,3..topx],
                                                  y <- [2,4..topy]],
		              n `mod` 60 `elem` [7,19,31,43]])
                    sieve


thirdStep size sieve =
	  let topx = floor(sqrt(fromIntegral size)) in
          flipAll (sort [n | n <- [3*x^2 - y^2 | x <- [1..topx],
                                                 y <- [(x-1),(x-3)..1],
						 x > y],
			     n `mod` 60 `elem` [11,23,47,59]])
                  sieve


unmarkMultiples limit n sieve =
    		let nonPrimes = [y | y <-[n,n+n..limit]] in
    		unmarkAll nonPrimes sieve


unmarkAll _        []         = []
unmarkAll []       sieve      = sieve
unmarkAll (np:nps) ((s,b):ss) 
	  | np == s	     = unmarkAll nps ss
	  | np < s	     = unmarkAll nps ((s,b):ss)
	  | otherwise	     = (s,b) : (unmarkAll (np:nps) ss)


atkin1 limit =
       aSieve1 limit
       	       ((thirdStep limit) .
	        (secondStep limit) .
		(firstStep limit) $
		initialAtkinSieve limit)
	       [(5,1),(3,1),(2,1)]


aSieve1 _     []         primes = primes
aSieve1 limit ((x,b):xs) primes
	       | b == 1		= aSieve1 limit (unmarkMultiples limit (x^2) xs) ((x,b): primes)
	       | otherwise	= aSieve1 limit xs primes


